### PR TITLE
[uCAIP] Midpoint demo update

### PR DESCRIPTION
**Pull request by shunr**
- Model view and evaluation
- Fixes to dataset importing from Bigquery, GCS, file sources

Contains changes in https://github.com/GoogleCloudPlatform/jupyter-extensions/pull/89

### CODE REVIEW

File: jupyterlab_automl/jupyterlab_automl/service.py [hunk 1]
```diff
@@ -1,284 +1,247 @@
...
+      if 'image' in model.metadata_schema_uri:
+        model_type = 'IMAGE'
+      elif 'tables' in model.metadata_schema_uri:
+        model_type = 'TABLE'
+      models.append({
+            "id": model.name,
+            "displayName": model.display_name,
+            "pipelineId": model.training_pipeline,
+            "createTime": (model.create_time.year, model.create_time.month, model.create_time.day, model.create_time.hour, model.create_time.minute, model.create_time.second),
+            "updateTime": (model.update_time.year, model.update_time.month, model.update_time.day, model.update_time.hour, model.update_time.minute, model.update_time.second),
```
Comment (josie-garza) [inline jupyterlab_automl/jupyterlab_automl/service.py:102]: changes times to timestamps to serialize and create Date objects in the TypeScript

File: jupyterlab_automl/src/components/model_properties.tsx [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+  async componentDidMount() {
+    this.getPipeline();
+  }
+
+  private getBudget(milliNodeHours: number): string {
+    return (milliNodeHours / 1000).toString() + ' node hours';
+  }
+
+  private getElapsedTime(totalSeconds: number): string {
```
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/src/components/model_properties.tsx:159]: We removed this in a subsequent pr and replaced with the humanizeDuration library

File: jupyterlab_automl/src/components/model_properties.tsx [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+      const seconds = totalSeconds - minutes * 60;
+      return minutes + ' min ' + seconds + ' sec';
+    }
+  }
+
+  private createData(key: string, val: any) {
+    return { key, val };
+  }
+
+  private getModelDetails(pipeline: Pipeline): any[] {
```
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/src/components/model_properties.tsx:176]: also refactored this in the next PR, but will definitely try change that to use this pattern

File: jupyterlab_automl/jupyterlab_automl/service.py [hunk 4]
```diff
@@ -1,284 +1,243 @@
...
+    metrics = []
+    for confidence_metric in confidence_metrics:
+      metric = {}
+      for label in labels:
+        if label == "confidenceThreshold":
+          value = confidence_metric.get(label, 0.0) * 100
         else:
-            return []
+          value = confidence_metric.get(label, "NaN")
+        metric[label] = value
```
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/jupyterlab_automl/service.py:137]:
```        
if label == "confidenceThreshold":
  value = confidence_metric.get(label, 0.0) * 100
else:
  value = confidence_metric.get(label, "NaN")
metric[label] = value
```
The code looks like this - it looks like we're setting the value variable on each iteration with a different default based on the label value. Let me know if this makes more sense

File: jupyterlab_automl/src/components/model_properties.tsx [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+interface RowProps {
+  row: any;
+  transformationOptions: any[];
+}
+
+interface RowState {
+  open: boolean;
+}
+
+const properties = [
```
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/src/components/model_properties.tsx:39]: This file got heavily refactored, but will make this change in the next PR

File: jupyterlab_automl/jupyterlab_automl/service.py [hunk 6]
```diff
@@ -1,284 +1,243 @@
...
+      metrics.append(metric)
+    return metrics
+
+  def get_model_evaluation(self, model_id):
```
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/jupyterlab_automl/service.py:141]: I can see where you're coming from, the function gets data from a few places. Personally, I prefer not creating single use helper functions since having to jump to another place in the code interrupts the logical flow, especially here where get_model_evaluation is already a pure fn

File: jupyterlab_automl/jupyterlab_automl/service.py [hunk 7]
```diff
@@ -1,284 +1,243 @@
...
+            "columnName": column[data_type]["columnName"]
+        })
+    return transformations
+
+  def get_pipeline(self, pipeline_id):
+    pipeline = self._pipeline_client.get_training_pipeline(name=pipeline_id)
+    optional_fields = ["targetColumn", "predictionType", "optimizationObjective", "budgetMilliNodeHours", "trainBudgetMilliNodeHours"]
+    training_task_inputs = json_format.MessageToDict(
+        pipeline._pb.training_task_inputs)
+    training_pipeline = {
```
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/jupyterlab_automl/service.py:172]: I'm not too sure what that looks like in this context - could you clarify what a translation function means?

File: jupyterlab_automl/src/components/list_resources_panel.tsx [hunk 8]
```diff
@@ -34,6 +35,9 @@ interface State {
   models: Model[];
   resourceType: ResourceType;
   searchString: string;
+  deleteDialog: boolean;
+  deleteSubmit: () => void;
+  deleteString: string;
```
[the remark this answers is not in the document]
Comment (josie-garza) [reply jupyterlab_automl/src/components/list_resources_panel.tsx:45]: Agreed. This component has a lot of state variables and I am wondering if some of them can be made local.

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (josie-garza) [reply jupyterlab_automl/jupyterlab_automl/service.py:143]: I think the reasoning here is because the confidenceThreshold needs to have a default as a number instead of a string because of the logic in the TypeScript

File: jupyterlab_automl/jupyterlab_automl/service.py [hunk 9]
```diff
@@ -1,284 +1,243 @@
...
+        model_type = 'TABLE'
+      models.append({
+            "id": model.name,
+            "displayName": model.display_name,
+            "pipelineId": model.training_pipeline,
+            "createTime": get_milli_time(model.create_time),
+            "updateTime": get_milli_time(model.update_time),
+            "etag": model.etag,
+            "modelType": model_type
+        })
```
[the remark this answers is not in the document]
Comment (josie-garza) [reply jupyterlab_automl/jupyterlab_automl/service.py]: yes I added this in the newest commit

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (josie-garza) [reply jupyterlab_automl/jupyterlab_automl/service.py:180]: is it something like `training_pipeline = self._get_base_pipeline(pipeline)` ? like abstracting it out into a helper function maybe

File: jupyterlab_automl/src/components/list_resources_panel.tsx [hunk 10]
```diff
@@ -334,12 +428,20 @@ export class ListResourcesPanel extends React.Component<Props, State> {
   }
 
   private async getDatasets() {
-    const datasets = await DatasetService.listDatasets();
-    this.setState({ datasets: datasets });
+    try {
+      const datasets = await DatasetService.listDatasets();
+      this.setState({ datasets: datasets });
+    } catch (e) {
+      console.log('Failed to load dataset resource.');
```
[the remark this answers is not in the document]
Comment (josie-garza) [reply jupyterlab_automl/src/components/list_resources_panel.tsx:436]: This error would only show if the user didn't have the API installed. But now that we've added checking to make sure the API is enabled we can probably take this out because the user would never get here anyways.

File: jupyterlab_automl/src/components/model_evaluation.tsx [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+    const marks = [];
+    for (let i = 0; i < modelEvaluation.confidenceMetrics.length; i++) {
+      marks.push({
+        value: modelEvaluation.confidenceMetrics[i].confidenceThreshold,
+      });
+    }
+    return marks;
+  }
+
+  private updateEvaluationTable(
```
[the remark this answers is not in the document]
Comment (josie-garza) [reply jupyterlab_automl/src/components/model_evaluation.tsx:329]: I think abstracting the adding the entries from the confidenceMetric would be a good way to break this up. Basically everything after the `if metric`

File: jupyterlab_automl/src/components/list_resources_panel.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (timothyrjames) [reply jupyterlab_automl/src/components/list_resources_panel.tsx:436]: I think you could leave it in, but maybe add a comment on why the user doesn't need to be alerted. I like to see comments used sparingly, but they can be great for explaining *why* you're doing something.

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (timothyrjames) [reply jupyterlab_automl/jupyterlab_automl/service.py:143]: I see - so given that we're doing the same thing for all cases except "confidenceThreshold", this pattern might be more readable (and a slight bit more efficient):

```Python
for label in labels:
  metric[label] = confidence_metric.get(label, "NaN")
metric["confidenceThreshold"] = confidence_metric.get(label, 0.0) * 100
```

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (timothyrjames) [reply jupyterlab_automl/jupyterlab_automl/service.py:147]: I get your point, but I'm curious to hear other proposals on how to simplify / make this more understandable as "get" implies something simple and this, as it stands, is quite a long method - hard to follow for a simple get method.

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (timothyrjames) [reply jupyterlab_automl/jupyterlab_automl/service.py:180]: I'm thinking something like:

```Python
def build_pipeline_dict(pipeline):
  return {
        "id": pipeline.name,
        "displayName": pipeline.display_name,
        "createTime": get_milli_time(pipeline.create_time),
        "updateTime": get_milli_time(pipeline.update_time),
        "elapsedTime": (pipeline.end_time - pipeline.start_time).seconds,
        "datasetId": pipeline.input_data_config.dataset_id,
  }
```

then it's simple and easily readable (and more concise) to see:

```Python
training_pipeline = build_pipeline_dict(pipeline)
```

File: jupyterlab_automl/src/components/model_evaluation.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (timothyrjames) [reply jupyterlab_automl/src/components/model_evaluation.tsx:329]: sg

File: jupyterlab_automl/src/components/list_resources_panel.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/src/components/list_resources_panel.tsx]: I renamed the variables in the latest commit, but maybe we can refactor the delete modal to use fewer state variables

File: jupyterlab_automl/src/components/list_resources_panel.tsx [hunk 12]
```diff
@@ -334,12 +428,20 @@ export class ListResourcesPanel extends React.Component<Props, State> {
...
   }
 
   private async getModels() {
-    const models = await ModelService.listModels();
-    this.setState({ models: models });
+    try {
+      const models = await ModelService.listModels();
+      this.setState({ models: models });
+    } catch (e) {
+      console.log('Failed to load models resource.');
```
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/src/components/list_resources_panel.tsx:445]: Will add a TODO, currently haven't thought of a clear/clean UX way to show this kind of error

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/jupyterlab_automl/service.py:186]: Ah, that makes sense - I think we do a similar refactor in #124 so this can be marked resolved for now

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/jupyterlab_automl/service.py:146]: one problem with this approach is that the confidenceThreshold field might not always exist on the confidence_metric - for example on regression models

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/jupyterlab_automl/service.py:150]: maybe separating out the code for parsing the fields could be a good idea

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (josie-garza) [reply jupyterlab_automl/jupyterlab_automl/service.py:146]: I think this function won't even be called for regression models because of ```    if "confidenceMetrics" in metrics:
      model_eval["confidenceMetrics"] = self._get_confidence_metrics(metrics["confidenceMetrics"])```

File: jupyterlab_automl/src/components/model_properties.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/src/components/model_properties.tsx:176]: But renaming a few private _get methods to _build would make a lot of sense

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shunr) [reply jupyterlab_automl/jupyterlab_automl/service.py:146]: oh nevermind that makes sense!

File: jupyterlab_automl/jupyterlab_automl/service.py [hunk 13]
```diff
@@ -1,284 +1,243 @@
...
 from enum import Enum
 
-from google.cloud import automl_v1beta1
-
+from google.cloud import aiplatform_v1alpha1, exceptions, storage
+from google.protobuf import json_format
+from googleapiclient.discovery import build
 from gcp_jupyterlab_shared.handlers import AuthProvider
 
+API_ENDPOINT = "us-central1-aiplatform.googleapis.com"
```
[the remark this answers is not in the document]
Comment (josie-garza) [reply jupyterlab_automl/jupyterlab_automl/service.py:14]: test

File: jupyterlab_automl/jupyterlab_automl/service.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (timothyrjames) [reply jupyterlab_automl/jupyterlab_automl/service.py:146]: Also, ```labels``` is a predefined list in the method, so either way I think they're logically equivalent.
